### DISCUSSION TITLE
Specific zone with depth of field

### DISCUSSION THREAD

**Discussion by brigidatm** [RayTracing/raytracing.github.io#963] [Q&A]
I replicated the code from the second book. I would like to focus on a single object of the scene, as was the case in the first book with the three spheres.
![example1](https://user-images.githubusercontent.com/20044838/140775452-87c542af-0d40-4cc9-aca2-1af02e1ec22f.png)

How do I focus on a single area like the one highlighted in yellow in the following figure?
![example](https://user-images.githubusercontent.com/20044838/140777986-1a299695-3eca-43d0-99ba-f12eb5652218.png)

the relative code is:
` const auto aspect_ratio = 1.0;
const int image_width = 400;
const int image_height = static_cast<int>(image_width / aspect_ratio);
const int samples_per_pixel = 100;
const int max_depth = 10;

point3 lookfrom(480, 250, -600);
point3 lookat(200, 280, 300);
vec3 vup(0, 1, 0);
double vfov = 40;
auto dist_to_focus = (lookfrom - lookat).length();
auto aperture = 0.2; `

thank you for the support, best regards.

**Comment by hollasch** [comment]
As presented above, your focus distance is derived from the look-from and look-at points. However, you don't need to make your focus distance dependent on these two parameters. Instead, set the focus distance independently to the distance between the camera "from" point and the focus plane (parallel to the image plane) that you want. If you don't want to go through the math, you can zero in iteratively with different focus distances (like fiddling with the focus ring on your physical camera).

**Comment by brigidatm** [comment]
if i use this code: 
`const auto aspect_ratio = 1.0;
const int image_width = 400;
const int image_height = static_cast<int>(image_width / aspect_ratio);
const int samples_per_pixel = 30;
const int max_depth = 10;

point3 lookfrom(480, 250, -600);
point3 lookat(200, 280, 300);
vec3 vup(0, 1, 0);
double vfov = 40;
**auto dist_to_focus = 20.0**;
auto aperture = 0.2;`
 
I don't get the desired result because the whole scene is blurred. I want only one object to be in focus with the rest of the scene out of focus. what am I doing wrong?

**Reply by hollasch** [reply]
You have to set the distance to the image plane where everything is in focus.

![dist-to-focus](https://user-images.githubusercontent.com/313871/141057930-044596a8-b997-4f91-89f7-69f72dac47db.jpg)

Here, you know your look-from point (480,250,-600) and your look-at point (200,280,300). You also must pick some point on (or near enough) the surface of the object you want in focus (labeled *O* in the figure). The image plane is perpendicular to the view vector from `lookfrom` to `lookat`, and the vector from the projection point `lookfrom` to *O* is just *O* - `lookfrom`. The length of that vector gives us *b*, the distance from the projection point to the point you want in focus `(O - lookfrom).length()`.

Since the image plane is perpendicular to line of sight, we can use simple trig to find *d*, the focus distance. To find cosθ, you just need to find the cosine of the angle between the two vectors.

For two vectors *A* and *B*,

*A* ⋅ *B* = |*A*| |*B*| cos(t)

so

cos(t) = (*A* ⋅ *B*) / (|*A*| |*B*|) = `dot(A,B) / (A.length() * B.length())`

In our case, A = `lookat - lookfrom`, and B = `O - lookfrom`.

**Reply by hollasch** [reply]
I think the following should work.

    point3 focus_point(?,?,?);
    auto focus_vec = focus_point - lookfrom;
    auto view_vec = lookat - lookfrom;
    auto b = focus_vec.length();
    auto cos_theta = dot(focus_vec, view_vec) / (focus_vec.length() * view_vec.length());
    auto dist_to_focus = b / cos_theta;

**Reply by brigidatm** [reply]
Hi @hollasch ,

thanks for the quick reply.

I just tried your suggested code but without success. 
Probably something in my code is wrong or I failed in setting points (es. focus_point).

Below the code used:

```
auto world = final_scene();
point3 lookfrom(478, 278, -600);
point3 lookat(278, 278, 0);
vec3 vup(0, 1, 0);
double vfov = 40;
auto aperture = 1;
```
```
point3 focus_point(200, 200, 0);
auto focus_vec = focus_point - lookfrom;
auto view_vec = lookat - lookfrom;
auto b = focus_vec.length();
auto cos_theta = dot(focus_vec, view_vec)/ (focus_vec.length() * view_vec.length());
auto dist_to_focus = b / cos_theta;
printf("dist_to_focus: %lf \n", dist_to_focus);

camera cam(lookfrom, lookat, vup, vfov, aspect_ratio, aperture, dist_to_focus, 0.0, 1.0);
```

The result is the following image (**there seems to be no blur in the area previously highlighted with the yellow box.**)

![image](https://user-images.githubusercontent.com/20044838/141165508-77b5df94-1706-468c-b849-c8f2392a5f28.png)

This is the code about the scene:
```
hittable_list final_scene() {
  hittable_list boxes1;
  auto ground = make_shared<lambertian>(color(0.48, 0.83, 0.53));

  const int boxes_per_side = 20;
  for (int i = 0; i < boxes_per_side; i++) {
    for (int j = 0; j < boxes_per_side; j++) {
      auto w = 100.0;
      auto x0 = -1000.0 + i * w;
      auto z0 = -1000.0 + j * w;
      auto y0 = 0.0;
      auto x1 = x0 + w;
      auto y1 = random_double(1, 101);
      auto z1 = z0 + w;

      boxes1.add(make_shared<box>(point3(x0, y0, z0), point3(x1, y1, z1), ground));
    }
  }

  hittable_list objects;

  objects.add(make_shared<bvh_node>(boxes1, 0, 1));

  auto light = make_shared<diffuse_light>(color(7, 7, 7));
  objects.add(make_shared<xz_rect>(123, 423, 147, 412, 554, light));
 
  hittable_list boxes2;
  auto white = make_shared<lambertian>(color(.73, .73, .73));
  int ns = 1000;
  for (int j = 0; j < ns; j++) {
    boxes2.add(make_shared<sphere>(point3::random(0, 165), 10, white));
  }

  objects.add(make_shared<translate>(
    make_shared<rotate_y>(
      make_shared<bvh_node>(boxes2, 0.0, 1.0), 15),
    vec3(-100, 270, 395)
    )
  );

  return objects;
}
```

Thank you for your availability,
Best regards

**Reply by brigidatm** [reply]
I tested his code on different scenes and I noticed that in the scenes where the lookfrom and the lookat have small values ​​and therefore the dist to focus is small, I have the desired focus.
an example:
`point3 lookfrom(20, 10, 14);
point3 lookat(0, 2, 0);
auto dist_to_focus = b / cos_theta (= 15)`

in the scene shown above or in the scene with the cornell box of the book (where the value of dist to focus is high, greater than 100), its code does not work. what am I doing wrong?
Thank you for your availability.

**Comment by hollasch** [comment]
Ah, I made a mistake. cosθ = adjacent / hypotenuse = d/b. So

    auto dist_to_focus = b * cos_theta;

**Reply by brigidatm** [reply]
Excuse me if I disturb you again but the code still doesn't work for the two cornell box scenes in the book. why this? what can i do to solve the problem? I await your suggestion, thanks

**Comment by hollasch** [comment]
Well, you could try reverse-engineering it. Did you try a search for values to see which one gets you the result you're looking for?

**Reply by brigidatm** [reply]
I did different test with different for dist_to_focus(manually chosen). Below a few example:

![photo5811945884227844041](https://user-images.githubusercontent.com/20044838/142239063-06aa7bf3-15d3-4b0f-ad43-daa73bd9b95a.jpg)

```
const auto aspect_ratio = 1.0;
const int image_width = 600;
const int image_height = static_cast<int>(image_width / aspect_ratio);
const int samples_per_pixel = 10;
const int max_depth = 5;
const color background = color(0.70, 0.80, 1.00);
```

As you can see, if I increase the value of dist_to_focus I get a little blurry scene.
the problem is that the blur affects the ENTIRE scene, I want to focus on a single area
